lib/cli-engine/cascading-config-array-factory.js

```diff
@@ -0,0 +1,391 @@
+     */
+    getConfigArrayForFile(filePath = "a.js") {
+        const { cwd } = internalSlotsMap.get(this);
+        const directoryPath = path.dirname(path.resolve(cwd, filePath));
+
+        debug(`Load config files for ${directoryPath}.`);
+
+        return this._finalizeConfigArray(
+            this._loadConfigInAncestors(directoryPath),
+            directoryPath
+        );
+    }
+
+    /**
+     * Clear config cache.
+     * @returns {void}
+     */
+    clearCache() {
+        const slots = internalSlotsMap.get(this);
+
+        slots.baseConfigArray = createBaseConfigArray(slots);
+        slots.cliConfigArray = createCLIConfigArray(slots);
+        slots.configCache.clear();
+    }
+
+    /**
+     * Load and normalize config files from the ancestor directories.
+     * @param {string} directoryPath The path to a leaf directory.
+     * @returns {ConfigArray} The loaded config.
+     * @private
+     */
+    _loadConfigInAncestors(directoryPath) {
+        const {
+            baseConfigArray,
+            configArrayFactory,
+            configCache,
+            cwd,
+            useEslintrc
+        } = internalSlotsMap.get(this);
+
+        if (!useEslintrc) {
+            return baseConfigArray;
+        }
+
+        let configArray = configCache.get(directoryPath);
+
+        // Hit cache.
+        if (configArray) {
+            debug(`Cache hit: ${directoryPath}.`);
+            return configArray;
+        }
+        debug(`No cache found: ${directoryPath}.`);
+
+        const homePath = os.homedir();
+
+        // Consider this is root.
+        if (directoryPath === homePath && cwd !== homePath) {
+            debug("Stop traversing because of considered root.");
+            configCache.set(directoryPath, baseConfigArray);
+            return baseConfigArray;
+        }
+
+        // Load the config on this directory.
+        try {
+            configArray = configArrayFactory.loadOnDirectory(directoryPath);
+        } catch (error) {
+            /* istanbul ignore next */
+            if (error.code === "EACCES") {
+                debug("Stop traversing because of 'EACCES' error.");
+                configCache.set(directoryPath, baseConfigArray);
+                return baseConfigArray;
+            }
+            throw error;
+        }
+
+        if (configArray.length > 0 && configArray.root) {
+            debug("Stop traversing because of 'root:true'.");
+            configCache.set(directoryPath, configArray);
+            return configArray;
+        }
+
+        // Load from the ancestors and merge it.
+        const parentPath = path.dirname(directoryPath);
+        const parentConfigArray = parentPath && parentPath !== directoryPath
+            ? this._loadConfigInAncestors(parentPath)
+            : baseConfigArray;
+
+        if (configArray.length > 0) {
+            configArray.unshift(...parentConfigArray);
+        } else {
+            configArray = parentConfigArray;
```

Comment:
Indeed, because `ConfigArray` inherits `Array`, it's mutable. But `ConfigArray#extractConfig()` method makes cache, so mutation after `ConfigArray#extractConfig()` was called causes bugs. Currently, `ConfigArray` is private and mutation happens only in initialization.

Probably we should freeze those for our maintenance.